pkg/util/ip_reservation_test.go

```diff
@@ -0,0 +1,183 @@
+	}
+}
+
+func TestValidateCIDR(t *testing.T) {
+	testCIDR := "192.168.92.0/29"
+	ipAllocator := initTestIPAllocator()
+	err := ipAllocator.ValidateCIDR(testCIDR)
+	if err != nil {
+		t.Errorf(err.Error())
+	}
+
+	testCIDR = "192.168.92.0"
+	err = ipAllocator.ValidateCIDR(testCIDR)
+	if err == nil {
+		t.Errorf("Invalid CIDR %s evaluated to be valid", testCIDR)
+	}
+}
+
+func TestGetUnReservedIP(t *testing.T) {
+	ips := [8]string{"192.168.92.3/29", "192.168.92.10/29", "192.168.92.20/29", "192.168.92.28/29"} // Using IPs which are not the beginning IPs of /29 CIDRs to evaluate the edge case
+	cidr := "192.168.92.0/27"
+	// Reserved IPV4 ranges listed by the cloud provider
+	reservedIPV4Ranges := make(map[string]bool)
+	cases := []struct {
+		name          string
+		expected      string
+		errorExpected bool
+	}{
+		{
+			name:          "0/4 /29 IP blocks used",
+			expected:      "192.168.92.0/29",
+			errorExpected: false,
+		}, {
+			name:          "1/4 /29 IP blocks used",
+			expected:      "192.168.92.8/29",
+			errorExpected: false,
+		},
+		{
+			name:          "2/4 /29 IP blocks used",
+			expected:      "192.168.92.16/29",
+			errorExpected: false,
+		},
+		{
+			name:          "3/4 /29 IP blocks used",
+			expected:      "192.168.92.24/29",
+			errorExpected: false,
+		},
+		{
+			name:          "All /29 IP blocks used",
+			expected:      "",
+			errorExpected: true,
+		},
+	}
+
+	for index, test := range cases {
+		ipAllocator := initTestIPAllocator()
+		for i := 0; i < index; i++ {
+			// Distributing IPs between the reserved lists at the cloud provider end and the reserved list in the ipAllocator
+			if i%2 == 0 {
+				ipAllocator.reserved[ips[i]] = true
+			} else {
+				reservedIPV4Ranges[ips[i]] = true
+			}
+
+		}
+
+		ipBlock, err := ipAllocator.GetUnreservedIPBlock(cidr, reservedIPV4Ranges)
+		if err != nil && !test.errorExpected {
+			t.Errorf("test %q failed: got error %s, expected %s", test.name, err.Error(), test.expected)
+		} else if err == nil && test.errorExpected {
+			t.Errorf("test %q failed: got reserved IP range %s, expected error", test.name, ipBlock)
+		} else if ipBlock != test.expected {
+			t.Errorf("test %q failed: got reserved IP range %s, expected %s", test.name, ipBlock, test.expected)
+		}
+	}
+}
+func TestValidateCIDROverlap(t *testing.T) {
+	cases := []struct {
+		name     string
+		cidr1    string
+		cidr2    string
+		expected bool
+	}{
+		{
+			name:     "Overlapping CIDRs",
+			cidr1:    "192.168.92.0/29",
+			cidr2:    "192.168.92.48/26",
+			expected: true,
+		},
+		{
+			name:     "Non overlapping CIDRs",
+			cidr1:    "192.168.92.0/29",
+			cidr2:    "192.168.22.67/26",
+			expected: false,
+		},
+	}
+
+	for _, test := range cases {
+		overlap := isOverlap(test.cidr1, test.cidr2)
+		if overlap != test.expected {
+			t.Errorf("test %q failed: got CIDR overlap  as %t, expected %t", test.name, overlap, test.expected)
+		}
+	}
+}
+
+func TestIncrementIP(t *testing.T) {
+	step := byte(8)
```

Comment:
add tests for different step sizes. The function accepts that as an argument and it looks like the function itself is assuming that the step size is 8.